spark3/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java

```diff
@@ -0,0 +1,502 @@
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+public final class TestStructuredStreamingRead3 {
+  private static final Configuration CONF = new Configuration();
+  private static final Schema SCHEMA = new Schema(
+          optional(1, "id", Types.IntegerType.get()),
+          optional(2, "data", Types.StringType.get())
+  );
+  private static SparkSession spark = null;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Rule
+  public ExpectedException exceptionRule = ExpectedException.none();
+
+  @BeforeClass
+  public static void startSpark() {
+    TestStructuredStreamingRead3.spark = SparkSession.builder()
+            .master("local[2]")
+            .config("spark.sql.shuffle.partitions", 4)
+            .getOrCreate();
+  }
+
+  @AfterClass
+  public static void stopSpark() {
+    SparkSession currentSpark = TestStructuredStreamingRead3.spark;
+    TestStructuredStreamingRead3.spark = null;
+    currentSpark.stop();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testReadStreamOnIcebergTableWithMultipleSnapshots() throws IOException, TimeoutException {
+    File parent = temp.newFolder("parent");
+    File location = new File(parent, "test-table");
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).bucket("id", 3).build();
+    Table table = tables.create(SCHEMA, spec, location.toString());
+
+    List<List<SimpleRecord>> expected = getTestDataForMultipleSnapshots();
+    appendData(expected, location);
+
+    table.refresh();
+
+    try {
+      Dataset<Row> df = spark.readStream()
+              .format("iceberg")
+              .load(location.toString());
+      List<SimpleRecord> actual = processStreamTillEnd(df);
+
+      Assert.assertEquals(
+          expected.stream().flatMap(List::stream).collect(Collectors.toList()),
+          actual);
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testResumingStreamReadFromCheckpoint() throws IOException, TimeoutException, StreamingQueryException {
+    File parent = temp.newFolder("parent");
+    File location = new File(parent, "test-table");
+    File writerCheckpoint = new File(parent, "writer-checkpoint");
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).bucket("id", 3).build();
+    Table table = tables.create(SCHEMA, spec, location.toString());
+    final String tempView = "microBatchView";
+
+    try {
+      Dataset<Row> df = spark.readStream()
+          .format("iceberg")
+          .load(location.toString());
+      DataStreamWriter<Row> singleBatchWriter = df.writeStream()
+          .trigger(Trigger.Once())
```

Comment:
I think that the intent is that `Trigger.Once()` combined with `awaitTermination` in `processMicroBatch` will stop the stream, is that correct? I think that it would be helpful to have a comment above this explaining how this is causing Spark to checkpoint and start new streams for each call to `processMicroBatch`.